In [1]:
import numpy as np

### Question 1: 
Version of NumPy

In [2]:
np.__version__

'1.21.5'

In [3]:
import pandas as pd

In [4]:
# read data.txt as dataset 

dataset = pd.read_csv('data.txt')

In [5]:
dataset

Make       Model  Year                Engine Fuel Type  Engine HP  \
0          BMW  1 Series M  2011     premium unleaded (required)      335.0   
1          BMW    1 Series  2011     premium unleaded (required)      300.0   
2          BMW    1 Series  2011     premium unleaded (required)      300.0   
3          BMW    1 Series  2011     premium unleaded (required)      230.0   
4          BMW    1 Series  2011     premium unleaded (required)      230.0   
...        ...         ...   ...                             ...        ...   
11909    Acura         ZDX  2012     premium unleaded (required)      300.0   
11910    Acura         ZDX  2012     premium unleaded (required)      300.0   
11911    Acura         ZDX  2012     premium unleaded (required)      300.0   
11912    Acura         ZDX  2013  premium unleaded (recommended)      300.0   
11913  Lincoln      Zephyr  2006                regular unleaded      221.0   

       Engine Cylinders Transmission Type      Driven_Wheels  Number of Doors  \
0                   6.0            MANUAL   rear wheel drive              2.0   
1                   6.0            MANUAL   rear wheel drive              2.0   
2                   6.0            MANUAL   rear wheel drive              2.0   
3                   6.0            MANUAL   rear wheel drive              2.0   
4                   6.0            MANUAL   rear wheel drive              2.0   
...                 ...               ...                ...              ...   
11909               6.0         AUTOMATIC    all wheel drive              4.0   
11910               6.0         AUTOMATIC    all wheel drive              4.0   
11911               6.0         AUTOMATIC    all wheel drive              4.0   
11912               6.0         AUTOMATIC    all wheel drive              4.0   
11913               6.0         AUTOMATIC  front wheel drive              4.0   

                             Market Category Vehicle Size  Vehicle Style  \
0      Factory Tuner,Luxury,High-Performance      Compact          Coupe   
1                         Luxury,Performance      Compact    Convertible   
2                    Luxury,High-Performance      Compact          Coupe   
3                         Luxury,Performance      Compact          Coupe   
4                                     Luxury      Compact    Convertible   
...                                      ...          ...            ...   
11909             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11910             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11911             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11912             Crossover,Hatchback,Luxury      Midsize  4dr Hatchback   
11913                                 Luxury      Midsize          Sedan   

       highway MPG  city mpg  Popularity   MSRP  
0               26        19        3916  46135  
1               28        19        3916  40650  
2               28        20        3916  36350  
3               28        18        3916  29450  
4               28        18        3916  34500  
...            ...       ...         ...    ...  
11909           23        16         204  46120  
11910           23        16         204  56670  
11911           23        16         204  50620  
11912           23        16         204  50920  
11913           26        17          61  28995  

[11914 rows x 16 columns]

### Question 2: 
How many records are in the dataset?

In [6]:
dataset.shape[0]

11914

### Question 3: 
Who are the most frequent car manufacturers (top-3) according to the dataset?

In [7]:
df1 = dataset.groupby('Make').count()

In [8]:
df1.Model.sort_values(ascending = False)[:3]

Make
Chevrolet     1123
Ford           881
Volkswagen     809
Name: Model, dtype: int64

### Question 4: 
What's the number of unique Audi car models in the dataset?

In [9]:
dataset[dataset['Make'] == "Audi"].Model.nunique()

34

### Question 5: 
How many columns in the dataset have missing values?

In [10]:
dataset.isna().sum().sort_values()

Make                    0
Model                   0
Year                    0
Transmission Type       0
Driven_Wheels           0
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
Engine Fuel Type        3
Number of Doors         6
Engine Cylinders       30
Engine HP              69
Market Category      3742
dtype: int64

### Question 6:
1. Find the median value of "Engine Cylinders" column in the dataset.

In [11]:
dataset['Engine Cylinders'].median()

6.0

### Question 6:

2. Next, calculate the most frequent value of the same "Engine Cylinders".

In [12]:
dataset['Engine Cylinders'].mode()

0    4.0
Name: Engine Cylinders, dtype: float64

In [13]:
df2 = dataset.groupby('Engine Cylinders').count()
df2

Make  Model  Year  Engine Fuel Type  Engine HP  \
Engine Cylinders                                                   
0.0                 56     56    56                56         13   
3.0                 30     30    30                30         30   
4.0               4752   4752  4752              4752       4743   
5.0                225    225   225               225        225   
6.0               4489   4489  4489              4486       4473   
8.0               2031   2031  2031              2031       2031   
10.0                68     68    68                68         68   
12.0               230    230   230               230        230   
16.0                 3      3     3                 3          3   

                  Transmission Type  Driven_Wheels  Number of Doors  \
Engine Cylinders                                                      
0.0                              56             56               51   
3.0                              30             30               30   
4.0                            4752           4752             4752   
5.0                             225            225              225   
6.0                            4489           4489             4489   
8.0                            2031           2031             2031   
10.0                             68             68               68   
12.0                            230            230              229   
16.0                              3              3                3   

                  Market Category  Vehicle Size  Vehicle Style  highway MPG  \
Engine Cylinders                                                              
0.0                            51            56             56           56   
3.0                            26            30             30           30   
4.0                          3126          4752           4752         4752   
5.0                           170           225            225          225   
6.0                          2923          4489           4489         4489   
8.0                          1545          2031           2031         2031   
10.0                           68            68             68           68   
12.0                          230           230            230          230   
16.0                            3             3              3            3   

                  city mpg  Popularity  MSRP  
Engine Cylinders                              
0.0                     56          56    56  
3.0                     30          30    30  
4.0                   4752        4752  4752  
5.0                    225         225   225  
6.0                   4489        4489  4489  
8.0                   2031        2031  2031  
10.0                    68          68    68  
12.0                   230         230   230  
16.0                     3           3     3

Note that Engine Cylinders corresponding to 4 has the maximum value of 4752. Hence the answer is 4.

### Question 6:

3. Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.

In [14]:
dataset['Engine Cylinders'] = dataset['Engine Cylinders'].fillna(4)

In [15]:
dataset.groupby('Engine Cylinders').count()

Make  Model  Year  Engine Fuel Type  Engine HP  \
Engine Cylinders                                                   
0.0                 56     56    56                56         13   
3.0                 30     30    30                30         30   
4.0               4782   4782  4782              4782       4772   
5.0                225    225   225               225        225   
6.0               4489   4489  4489              4486       4473   
8.0               2031   2031  2031              2031       2031   
10.0                68     68    68                68         68   
12.0               230    230   230               230        230   
16.0                 3      3     3                 3          3   

                  Transmission Type  Driven_Wheels  Number of Doors  \
Engine Cylinders                                                      
0.0                              56             56               51   
3.0                              30             30               30   
4.0                            4782           4782             4782   
5.0                             225            225              225   
6.0                            4489           4489             4489   
8.0                            2031           2031             2031   
10.0                             68             68               68   
12.0                            230            230              229   
16.0                              3              3                3   

                  Market Category  Vehicle Size  Vehicle Style  highway MPG  \
Engine Cylinders                                                              
0.0                            51            56             56           56   
3.0                            26            30             30           30   
4.0                          3156          4782           4782         4782   
5.0                           170           225            225          225   
6.0                          2923          4489           4489         4489   
8.0                          1545          2031           2031         2031   
10.0                           68            68             68           68   
12.0                          230           230            230          230   
16.0                            3             3              3            3   

                  city mpg  Popularity  MSRP  
Engine Cylinders                              
0.0                     56          56    56  
3.0                     30          30    30  
4.0                   4782        4782  4782  
5.0                    225         225   225  
6.0                   4489        4489  4489  
8.0                   2031        2031  2031  
10.0                    68          68    68  
12.0                   230         230   230  
16.0                     3           3     3

Note the value corresponding to 4 Engine Cylinders has increased by 30 to 4782.

### Question 6:
4. Now, calculate the median value of "Engine Cylinders" once again.

In [16]:
dataset['Engine Cylinders'].median()

6.0

Has it changed?

Answer is No.

### Question 7:
1. Select all the "Lotus" cars from the dataset.

In [17]:
Lotus = dataset[dataset['Make'] == 'Lotus']
Lotus

Make      Model  Year                Engine Fuel Type  Engine HP  \
3912  Lotus      Elise  2009     premium unleaded (required)      189.0   
3913  Lotus      Elise  2009     premium unleaded (required)      218.0   
3914  Lotus      Elise  2009     premium unleaded (required)      189.0   
3915  Lotus      Elise  2010     premium unleaded (required)      189.0   
3916  Lotus      Elise  2010     premium unleaded (required)      218.0   
3917  Lotus      Elise  2011     premium unleaded (required)      189.0   
3918  Lotus      Elise  2011     premium unleaded (required)      217.0   
3919  Lotus      Elise  2011     premium unleaded (required)      217.0   
4216  Lotus     Esprit  2002     premium unleaded (required)      350.0   
4217  Lotus     Esprit  2003     premium unleaded (required)      350.0   
4218  Lotus     Esprit  2004     premium unleaded (required)      350.0   
4257  Lotus  Evora 400  2017     premium unleaded (required)      400.0   
4258  Lotus  Evora 400  2017     premium unleaded (required)      400.0   
4259  Lotus      Evora  2011     premium unleaded (required)      276.0   
4260  Lotus      Evora  2011     premium unleaded (required)      276.0   
4261  Lotus      Evora  2013     premium unleaded (required)      276.0   
4262  Lotus      Evora  2013     premium unleaded (required)      345.0   
4263  Lotus      Evora  2013     premium unleaded (required)      345.0   
4264  Lotus      Evora  2013     premium unleaded (required)      276.0   
4265  Lotus      Evora  2014     premium unleaded (required)      345.0   
4266  Lotus      Evora  2014     premium unleaded (required)      276.0   
4267  Lotus      Evora  2014     premium unleaded (required)      345.0   
4268  Lotus      Evora  2014     premium unleaded (required)      276.0   
4292  Lotus      Exige  2009  premium unleaded (recommended)      257.0   
4293  Lotus      Exige  2009  premium unleaded (recommended)      240.0   
4294  Lotus      Exige  2010  premium unleaded (recommended)      240.0   
4295  Lotus      Exige  2011  premium unleaded (recommended)      240.0   
4296  Lotus      Exige  2011  premium unleaded (recommended)      257.0   
4297  Lotus      Exige  2011  premium unleaded (recommended)      257.0   

      Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
3912               4.0            MANUAL  rear wheel drive              2.0   
3913               4.0            MANUAL  rear wheel drive              2.0   
3914               4.0            MANUAL  rear wheel drive              2.0   
3915               4.0            MANUAL  rear wheel drive              2.0   
3916               4.0            MANUAL  rear wheel drive              2.0   
3917               4.0            MANUAL  rear wheel drive              2.0   
3918               4.0            MANUAL  rear wheel drive              2.0   
3919               4.0            MANUAL  rear wheel drive              2.0   
4216               8.0            MANUAL  rear wheel drive              2.0   
4217               8.0            MANUAL  rear wheel drive              2.0   
4218               8.0            MANUAL  rear wheel drive              2.0   
4257               6.0            MANUAL  rear wheel drive              2.0   
4258               6.0            MANUAL  rear wheel drive              2.0   
4259               6.0            MANUAL  rear wheel drive              2.0   
4260               6.0            MANUAL  rear wheel drive              2.0   
4261               6.0            MANUAL  rear wheel drive              2.0   
4262               6.0            MANUAL  rear wheel drive              2.0   
4263               6.0            MANUAL  rear wheel drive              2.0   
4264               6.0            MANUAL  rear wheel drive              2.0   
4265               6.0            MANUAL  rear wheel drive              2.0   
4266               6.0            MANUAL  rear wheel drive              2.0   
4267              

2. Select only columns "Engine HP", "Engine Cylinders".

In [18]:
Lotus1 = Lotus[['Engine HP', 'Engine Cylinders']]
Lotus1

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3914      189.0               4.0
3915      189.0               4.0
3916      218.0               4.0
3917      189.0               4.0
3918      217.0               4.0
3919      217.0               4.0
4216      350.0               8.0
4217      350.0               8.0
4218      350.0               8.0
4257      400.0               6.0
4258      400.0               6.0
4259      276.0               6.0
4260      276.0               6.0
4261      276.0               6.0
4262      345.0               6.0
4263      345.0               6.0
4264      276.0               6.0
4265      345.0               6.0
4266      276.0               6.0
4267      345.0               6.0
4268      276.0               6.0
4292      257.0               4.0
4293      240.0               4.0
4294      240.0               4.0
4295      240.0               4.0
4296      257.0               4.0
4297      257.0               4.0

3. Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).

In [19]:
Lotus1 = Lotus1.drop_duplicates()
Lotus1

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3918      217.0               4.0
4216      350.0               8.0
4257      400.0               6.0
4259      276.0               6.0
4262      345.0               6.0
4292      257.0               4.0
4293      240.0               4.0

4. Get the underlying NumPy array. Let's call it X.

In [20]:
X = np.array(Lotus1)

In [21]:
X

array([[189.,   4.],
       [218.,   4.],
       [217.,   4.],
       [350.,   8.],
       [400.,   6.],
       [276.,   6.],
       [345.,   6.],
       [257.,   4.],
       [240.,   4.]])

5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.

In [22]:
X.T

array([[189., 218., 217., 350., 400., 276., 345., 257., 240.],
       [  4.,   4.,   4.,   8.,   6.,   6.,   6.,   4.,   4.]])

In [23]:
XTX = X.T.dot(X)

In [24]:
XTX

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

6. Invert XTX.

In [25]:
XTX_inv = np.linalg.inv(XTX)

In [26]:
XTX_inv

array([[ 5.53084235e-05, -2.94319825e-03],
       [-2.94319825e-03,  1.60588447e-01]])

7. Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].

In [27]:
y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])

In [28]:
y

array([1100,  800,  750,  850, 1300, 1000, 1000, 1300,  800])

8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.

In [29]:
w = XTX_inv.dot(X.T).dot(y)

In [30]:
w

array([  4.59494481, -63.56432501])

9. What's the value of the first element of w?

In [31]:
w[0]

4.594944810094579